File: exercises/kotlin/src/test/kotlin/com/adaptionsoft/games/trivia/runner/GoldenMaster.kt
package com.adaptionsoft.games.trivia.runner

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.PrintStream
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Paths

object GoldenMaster {
    private val BYTE_ENCODING: Charset = StandardCharsets.UTF_8

    fun getResult(i: Int): String {
        return try {
            String(
                Files.readAllBytes(
                    Paths.get("src", "test", "resources", String.format("output%d.txt", i))
                ), BYTE_ENCODING
            )
        } catch (e: IOException) {
            generateExpectedResult(i)
        }
    }

    private fun generateExpectedResult(i: Int): String {
        val originalOut = System.out
        return try {
            ByteArrayOutputStream().use { outContent ->
                PrintStream(outContent, true, BYTE_ENCODING.name()).use { interceptingStream ->
                    System.setOut(interceptingStream)

                    GameRunner.main(arrayOf(Integer.toString(i)))

                    Files.write(
                        Paths.get("src", "test", "resources", String.format("output%d.txt", i)),
                        outContent.toByteArray()
                    )
                    outContent.toString()
                }
            }
        } catch (e: IOException) {
            throw RuntimeException("Failed to generate the expected result for iteration $i", e)
        } finally {
            System.setOut(originalOut)
        }
    }
}

File: exercises/php/tests/trivia/GoldenMaster.php
<?php

namespace Tests;

use trivia\GameRunner;

class GoldenMaster
{
    const BYTE_ENCODING = 'UTF-8';

    public static function getResult($i)
    {
        $filePath = sprintf("resources/output%d.txt", $i);

        if (file_exists($filePath)) {
            return file_get_contents($filePath, false, null, 0);
        }

        return self::generateExpectedResult($i);
    }

    private static function generateExpectedResult($i)
    {
        // Your implementation to generate the expected result
        // ...

        $originalOut = fopen('php://output', 'wb');
        $outContent = fopen('php://temp', 'wb+');

        try {
            ob_start();
            try {
//                include 'GameRunner.php'; // Include or require the file containing GameRunner logic
                (new GameRunner())->runGame($i);
            } finally {
                $result = ob_get_clean();
                fwrite($outContent, $result);
                file_put_contents(__DIR__ . sprintf("/../../resources/output%d.txt", $i), $result);
                fclose($originalOut);
            }

            return stream_get_contents($outContent);
        } finally {
            fclose($outContent);
        }
    }
}
